pkg/quota/image/usage.go

```diff
@@ -10,139 +13,400 @@ import (
-	if len(image.DockerImageLayers) == 0 || image.DockerImageMetadata.Size == 0 {
-		if err = imageapi.ImageWithMetadata(image); err != nil {
-			glog.Errorf("Failed to parse metadata of image %q with DockerImageReference %q: %v", image.Name, image.DockerImageReference, err)
-			return nil, nil, err
-		}
-	}
-	isSize = resource.NewQuantity(0, resource.BinarySI)
-	if value, ok := image.Annotations[imageapi.ManagedByOpenShiftAnnotation]; ok && value == "true" {
-		sizeIncrement = resource.NewQuantity(image.DockerImageMetadata.Size, resource.BinarySI)
-	} else {
-		sizeIncrement = resource.NewQuantity(0, resource.BinarySI)
-	}
-	if is == nil {
+//  1. number of images currently tagged in the namespace; the image and images tagged in the given is don't
+//     count unless they are tagged in other is as well
+//  2. number of new internally managed images referenced either by the is or by the image
+//  3. an error if something goes wrong
+func (c *GenericImageStreamUsageComputer) GetProjectImagesUsageIncrement(
+	namespace string,
+	is *imageapi.ImageStream,
+	image *imageapi.Image,
+) (images, imagesIncrement *resource.Quantity, err error) {
+	processedImages := sets.NewString()
+
+	iss, err := c.listImageStreams(namespace)
+	if err != nil {
 		return
 	}
 
-	processedImages := make(map[string]sets.Empty)
-	processedLayers := make(map[string]sets.Empty)
+	imagesIncrement = resource.NewQuantity(0, resource.DecimalSI)
+
+	for _, imageStream := range iss.Items {
+		if is != nil && imageStream.Name == is.Name {
+			continue
+		}
+		c.processImageStreamImages(&imageStream, func(_, _ string, img *imageapi.Image) error {
+			processedImages.Insert(img.Name)
+			return nil
+		})
+	}
+
+	if is != nil {
+		c.processImageStreamImages(is, func(_, _ string, img *imageapi.Image) error {
+			if !processedImages.Has(img.Name) {
+				processedImages.Insert(img.Name)
+				imagesIncrement.Set(imagesIncrement.Value() + 1)
+			}
+			return nil
+		})
+	}
+
+	if image != nil && !processedImages.Has(image.Name) {
+		if value, ok := image.Annotations[imageapi.ManagedByOpenShiftAnnotation]; ok && value == "true" {
+			if !processedImages.Has(image.Name) {
+				imagesIncrement.Set(imagesIncrement.Value() + 1)
+			}
+		}
+	}
+
+	images = resource.NewQuantity(int64(len(processedImages)), resource.DecimalSI)
+
+	return
+}
 
-	for _, history := range is.Status.Tags {
+// processImageStreamImages is a utility method that calls a given handler for every image of the given image
+// stream that belongs to internal registry. It process image stream status and optionally spec.
+func (c *GenericImageStreamUsageComputer) processImageStreamImages(
+	is *imageapi.ImageStream,
+	handler func(tag string, dockerImageReference string, image *imageapi.Image) error,
+) error {
+	processedImages := sets.NewString()
+
+	for tag, history := range is.Status.Tags {
```

Comment:
This loop is poorly structured - it would be better to identify a full set of tags (unique) first, group them by their registry, and process them in order.  As is you will fetch the same image multiple times if it is tagged, potentially list the same image streams multiple times, etc.